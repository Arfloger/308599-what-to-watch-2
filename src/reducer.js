import {films} from './mocks/film';

const getFilms = (genre) => {

  if (genre === `All genres`) {
    return films.slice();
  }

  return films.slice().filter((it) => it.genre === genre);
};

export const initialState = {
  genre: `All genres`,
  films,
};

export const ActionCreator = {
  chengeGenre: (genre) => ({
    type: `CHENGE_GENRE`,
    payload: genre,
  }),

  getFilms: (genre) => ({
    type: `GET_FILMS`,
    payload: getFilms(genre),
  })
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case `CHENGE_GENRE`:
      return Object.assign({}, state, {
        genre: action.payload,
      });
    case `GET_FILMS`:
      return Object.assign({}, state, {
        films: action.payload,
      });
  }

  return state;
};
