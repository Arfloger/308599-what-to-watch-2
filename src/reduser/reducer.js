export const getGenreList = (films) => {

  let genresList = new Set(films.map((it) => it[`genre`]));
  initialState.genres = initialState.genres.concat(Array.from(genresList)).slice(0, 10);

  return initialState.genres;
};

export const filteredFilms = (genre) => {
  if (genre === `All genres`) {
    return initialState.films.slice();
  }

  return initialState.films.slice().filter((it) => it.genre === genre);
};

const MIN_CARDS_ON_PAGE = 8;

export const initialState = {
  genre: `All genres`,
  genres: [`All genres`],
  films: [],
  cardsOnPage: MIN_CARDS_ON_PAGE,
  requireAuthorization: false,
};

const ActionType = {
  CHANGE_GENRE: `CHANGE_GENRE`,
  LOAD_FILMS: `LOAD_FILMS`,
  FILTERED_FILMS: `FILTERED_FILMS`,
  GET_GENRE_LIST: `GET_GENRE_LIST`,
  INCREASE_QUANTITY_FILMS: `INCREASE_QUANTITY_FILMS`,
  RESET_TO_MIN_FILMS: `RESET_TO_MIN_FILMS`,
  REQUIRE_AUTHORIZATION: `REQUIRE_AUTHORIZATION`,
  LOGIN_USER: `LOGIN_USER`,
};

export const ActionCreator = {
  changeGenre: (genre) => ({
    type: ActionType.CHANGE_GENRE,
    payload: genre,
  }),

  loadFilms: (films) => ({
    type: ActionType.LOAD_FILMS,
    payload: films,
  }),

  filteredFilms: (genre) => ({
    type: ActionType.FILTERED_FILMS,
    payload: filteredFilms(genre),
  }),

  getGenreList: (genres) => ({
    type: ActionType.GET_GENRE_LIST,
    payload: getGenreList(genres),
  }),

  increaseQuantityFilms: () => ({
    type: ActionType.INCREASE_QUANTITY_FILMS,
    payload: 20,
  }),

  resetToMinFilms: () => ({
    type: ActionType.RESET_TO_MIN_FILMS,
    payload: MIN_CARDS_ON_PAGE,
  }),

  requireAuthorization: (status) => ({
    type: ActionType.REQUIRE_AUTHORIZATION,
    payload: status,
  }),

  loginUser: (data) => ({
    type: ActionType.LOGIN_USER,
    payload: data
  }),

};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      return Object.assign({}, state, {
        genre: action.payload,
      });
    case ActionType.LOAD_FILMS:
      return Object.assign({}, state, {
        films: action.payload,
      });
    case ActionType.GET_GENRE_LIST:
      return Object.assign({}, state, {
        genres: action.payload,
      });
    case ActionType.FILTERED_FILMS:
      return Object.assign({}, state, {
        films: action.payload,
      });
    case ActionType.INCREASE_QUANTITY_FILMS:
      return Object.assign({}, state, {
        cardsOnPage: state.cardsOnPage + action.payload,
      });
    case ActionType.RESET_TO_MIN_FILMS:
      return Object.assign({}, state, {
        cardsOnPage: action.payload,
      });
    case ActionType.REQUIRE_AUTHORIZATION:
      return Object.assign({}, state, {
        requireAuthorization: action.payload,
      });
    case ActionType.LOGIN_USER: return Object.assign({}, state, {
      id: action.payload.id,
      email: action.payload.email,
      name: action.payload.name,
      avatar: action.payload.avatar_url,
    });
  }

  return state;
};

export const Operation = {
  loadFilms: () => (dispatch, _getState, api) => {
    return api.get(`/films`)
      .then((response) => {
        dispatch(ActionCreator.loadFilms(response.data));
        dispatch(ActionCreator.getGenreList(response.data));
        initialState.films = response.data;
      });
  },
  checkAuth: (login, password) => {
    return (dispatch, _getState, api) => {
      return api
        .post(`/login`, {
          email: login,
          password,
        })
        .then((res) => {
          if (res.status >= 200) {
            dispatch(ActionCreator.requireAuthorization(true));
            dispatch(ActionCreator.loginUser(res.data));
          }
        })
        .catch((error) => {
          throw new Error(error);
        });
    };
  }
};
