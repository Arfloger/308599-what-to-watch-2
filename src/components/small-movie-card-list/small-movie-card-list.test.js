import React from "react";
import renderer from "react-test-renderer";
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {reducer} from "../../reduser/reducer";

import SmallMovieCardList from "./small-movie-card-list.jsx";

it(`SmallMovieCardList correctly renders after relaunch`, () => {
  const store = createStore(reducer);
  const tree = renderer.create(<Provider store={store}>
    <SmallMovieCardList films={[]}/></Provider>).toJSON();

  expect(tree).toMatchSnapshot();
});
