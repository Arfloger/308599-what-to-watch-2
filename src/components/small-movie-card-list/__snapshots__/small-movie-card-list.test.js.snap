// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallMovieCardList correctly renders after relaunch 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/the-grand-budapest-hotel-poster.jpg"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        The Grand Budapest Hotel
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/war-of-the-worlds.jpg"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        War of the worlds
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/we-need-to-talk-about-kevin.jpg"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        We need to talk about Kevin
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/macbeth.jpg"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Macbeth
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/johnny-english.jpg"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Johnny English
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/the-grand-budapest-hotel-poster.jpg"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        The Grand Budapest Hotel
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/war-of-the-worlds.jpg"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        War of the worlds
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/we-need-to-talk-about-kevin.jpg"
        src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        We need to talk about Kevin
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/macbeth.jpg"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Macbeth
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        height="175"
        muted={true}
        poster="img/johnny-english.jpg"
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Johnny English
      </a>
    </h3>
  </article>
</div>
`;
